### PR TITLE
[REVIEW] Adding cudf::sort and cudf::sort_by_key

### LINKED ISSUE #3580: [FEA] Add `cudf::sort`
**Is your feature request related to a problem? Please describe.**

There is currently no function to materialize the sorted order of a `table`. The only way to do this currently is to call `cudf::sorted_order` and then `cudf::gather`. 

**Describe the solution you'd like**

Add `cudf::sort` that is a fairly simple wrapper around `cudf::sorted_order` and `cudf::gather`. 

```c++
/**
 * @brief Performs a lexicographic sort of the rows of a table
 *
 * @param values The table to sort
 * @param column_order The desired order for each column. Size must be
 * equal to `input.num_columns()` or empty. If empty, all columns are sorted in
 * ascending order.
 * @param null_precedence The desired order of a null element compared to other
 * elements for each column in `keys`. Size must be equal to
 * `input.num_columns()` or empty. If empty, all columns will be sorted with
 * `null_order::BEFORE`.
 * @param mr The device memory resource used to allocate the returned table
 * @return New table containing the desired sorted order of `input`
 */
std::unique_ptr<table> sort(table_view const& input,
                            std::vector<order> const& column_order,
                            std::vector<null_order> const& null_precedence,
                            rmm::mr::device_memory_resource* mr);
```

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Adding cudf::sort and cudf::sort_by_key which will be using sorted_order and gather internally. Test cases have been updated to take care of these APIs as well.

close #3580

### CODE REVIEW

File: cpp/include/cudf/sorting.hpp [hunk 1]
```diff
@@ -63,5 +63,50 @@ bool is_sorted(cudf::table_view const& table,
...
 
+/**
+ * @brief Performs a lexicographic sort of the rows of a table
+ *
+ * @param values The table to sort
+ * @param column_order The desired order for each column. Size must be
+ * equal to `input.num_columns()` or empty. If empty, all columns are sorted in
+ * ascending order.
+ * @param null_precedence The desired order of a null element compared to other
+ * elements for each column in `keys`. Size must be equal to
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/sorting.hpp]: My mistake, it should be `input`

File: cpp/include/cudf/sorting.hpp [hunk 2]
```diff
@@ -63,5 +63,50 @@ bool is_sorted(cudf::table_view const& table,
...
+    std::vector<null_order> const& null_precedence = {},
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+
+/**
+ * @brief Performs a key-value sort.
+ *
+ * Creates a new table that reorders the rows of `values` according to the
+ * lexicographic ordering of the rows of `keys`.
+ *
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/sorting.hpp]: Missed that

File: cpp/src/sort/sort.cu [hunk 3]
```diff
@@ -87,6 +88,21 @@ std::unique_ptr<column> sorted_order(table_view input,
...
 }
+
+std::unique_ptr<table> sort_by_key(
+    table_view const& values, table_view const& keys,
+    std::vector<order> const& column_order,
+    std::vector<null_order> const& null_precedence,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream) {
+
+    CUDF_EXPECTS(values.num_rows() == keys.num_rows(), "Mismatch in number of rows for values and keys");
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/src/sort/sort.cu]: But, if you plan to use these `detail::` functions internally somewhere else you will not be handling these exceptions or you have to handle in both public and detail APIs, which is redundant.

[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/src/sort/sort.cu]: Ah... OK, yes, that make sense.

**Comment by rgsl888prabhu** [comment]
@trxcllnt

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**